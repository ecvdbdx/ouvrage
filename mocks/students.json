[
  {
    "username": "epatrois",
    "first_name": "Manu",
    "last_name": "Patrois",
    "id": "2"
  },
  {
    "id": "5",
    "username": "mdewaele",
    "first_name": "Maxence",
    "last_name": "De Waele",
    "email": "maxencedewaele@gmail.com",
    "birthdate": "838921463",
    "avatar": "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png"

  },
  {
    "id": "8",
    "username": "rodrigodelanoche",
    "first_name": "Vincent",
    "last_name": "Rodriguez",
    "email": "vincentrodrigu@hotmail.fr",
    "birthdate": "14/05/1413",
    "avatar": "http://etc.usf.edu/clipart/5500/5596/columbus_14_md.gif"
  },
  {
    "id": "6",
    "username": "erozis",
    "first_name": "Emilie",
    "last_name": "Rozis",
    "email": "emilie.40@hotmail.com"
  },
  {
    "first_name": "Hubert",
    "username": "hmoncenis",
    "id": "1",
    "last_name": "Montcenis"
  },
  {
    "id": "7",
    "username": "vivianeglz",
    "first_name": "Viviane",
    "last_name": "Gonzalez",
    "email": "viviane.gonzalez@outlook.fr",
    "birthdate": "02/06/1995",
    "avatar": "kikou.png"
  },
  {
    "id": "3",
    "username": "antoinesinety",
    "first_name": "Antoine",
    "last_name": "de Sinety",
    "email": "fake@yopmail.fr",
    "birthdate": "08/06/1995",
    "avatar": "Image.png"
  },
  {
    "id": "4",
    "username": "tgauzin",
    "first_name": "Thomas",
    "last_name": "Gauzin",
    "email": "thomas.gauzin@gmail.com"
  }
]
