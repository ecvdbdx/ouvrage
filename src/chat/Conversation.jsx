import React from 'react';

export default class Conversation extends React.Component {
  render () {
    const currentContact = this.props.currentContact;
    return (
      <div id="conversation">
        {
          currentContact ?
            <div className={currentContact === null ? 'hide' : ''}>
              <header>
                <h2>
                  { currentContact.username }
                </h2>
              </header>
              <section>
                <h3>
                  <div className="user-avatar" >
                    <img src={currentContact.avatar}/>
                  </div>
                  <div className="user-info">
                    <a>{currentContact.username}</a>
                  </div>
                </h3>
              </section>
            </div> : <div/>
        }
      </div>
    );
  }
}
