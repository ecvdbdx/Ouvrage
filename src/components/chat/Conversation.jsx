import React from 'react';
import Message from './Message.jsx';
import ChatApi from './ChatApi';

export default class Conversation extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      message: '',
      messages: [],
      autoScrollActive: false
    };
    this.submitMessage = this.submitMessage.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.scrollDown = this.scrollDown.bind(this);
    this.setAutoScroll = this.setAutoScroll.bind(this);

    this.refMessageList = React.createRef();

    if (this.props.currentContact) {
      this.retrieveConversation(this.props.currentContact, this.props.me);
      setInterval(() => {
        this.retrieveConversation(this.props.currentContact, this.props.me);
      }, 2000);
    }
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.currentContact !== this.props.currentContact)
      this.retrieveConversation(nextProps.currentContact, this.props.me);
  }

  retrieveConversation (contact, me) {
    ChatApi.getConversation(contact, me).then((messages) => {
      this.setState({
        messages: messages
      });
      this.scrollDown();
    });
  }

  handleChange (event) {
    this.setState({
      message: event.target.value
    });
  }


  submitMessage (event) {
    if (event.key === 'Enter') {
      const textField = event.target;
      const message = textField.value;
      const me = this.props.me;
      const user = this.props.currentContact;
      ChatApi.sendMessage(message, me, user).then((record) => {
        let messages = this.state.messages;
        messages.push(record.fields);
        this.setState({
          messages: messages,
          message: ''
        });
        textField.value = '';
        this.scrollDown();
      });
    }
  }

  scrollDown () {
    if (this.state.autoScrollActive) {
      const nodeElement = this.refMessageList.current;
      nodeElement.scrollTop = nodeElement.scrollHeight;
    }
  }

  setAutoScroll (event) {
    const nodeElement = this.refMessageList.current;
    if (nodeElement === null) return 0;
    const actuallScroll = nodeElement.scrollHeight - nodeElement.offsetHeight;
    if (actuallScroll - nodeElement.scrollTop <= 3) {
      this.setState({
        autoScrollActive: true
      });
    }
    else {
      this.setState({
        autoScrollActive: false
      });
    }
  }

  render () {
    const currentContact = this.props.currentContact;
    const me = this.props.me;
    if (!currentContact) return null;

    return (
      <section className="conversation" ref={this.refMessageList} onScroll={this.setAutoScroll}>
        {
          <div>
            <header>
              <h2>{currentContact.firstName} {currentContact.lastName}</h2>
              <img src={currentContact.avatar} className="user-avatar"/>
            </header>
            <main>
              <ul className="message-list">
                {
                  this.state.messages.map((message, index) => (
                    <Message key={index} message={message} me={me}/>
                  ))
                }
              </ul>
            </main>
            <footer className="message-section">
              <input type="text"
                placeholder="Votre message..."
                onChange={this.handleChange}
                value={this.state.message}
                onKeyPress={this.submitMessage}
              />
            </footer>
          </div>
        }
      </section>
    );
  }
}
