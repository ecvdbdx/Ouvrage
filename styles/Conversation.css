.conversation {
  border-bottom: 1px solid rgba(0, 0, 0, .05);
  min-height: 100vh;
  padding: 12px;
  position: relative;
  max-height: 100vh;
  overflow-y: scroll;
}

.conversation header{
  border-bottom: 1px solid rgba(0, 0, 0, .10);
  min-height: 70px;
  position: relative;
  margin-left: -12px;
  margin-right: -12px;
  display: flex;
  justify-content: center;
  align-items: center;
}

.conversation header .username {
  margin-left: 10px;
}

.conversation .user-avatar{
  background-color: rgba(0, 0, 0, .05);
  border-radius: 50%;
  overflow: hidden;
  height: 60px;
  width: 60px;
  position: absolute;
  top: 0;
  left: 10px;
}

.conversation .user-info{
  display: inline-block;
  vertical-align: top;
  padding-left: 10px;
  margin-left: 20px;
  line-height: 15px;
}
.conversation h3{
  display: inline-block;
  vertical-align: top;
  padding-left: 10px;
  margin-top: 5px;
}
.conversation .user-info a{
  color: #365899;
  cursor: pointer;
  text-decoration: none;
  font-size: 16px;
 }

.message-section {
  position: fixed;
  bottom: 0;
  width: 100%;
  display: flex;
  border-top: 1px solid rgba(0, 0, 0, .05);
  margin-left: -12px;
  margin-right: -12px;
}

.message-section input {
  border: 0;
  background: #fff;
  padding-left: 30px;
  min-height: 80px;
  font-size: 14px;
  outline: 0;
  width: 100%;
}

