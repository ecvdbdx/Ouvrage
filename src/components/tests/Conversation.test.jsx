import React from 'react';
import { shallow } from 'enzyme';

import Conversation from '../chat/Conversation.jsx';

import students from '../../../mocks/students.json';


it('renders correctly', () => {
  const wrapper = shallow(<Conversation currentContact={students[0]} me={students[1]} />);
  expect(wrapper).toMatchSnapshot();
});
