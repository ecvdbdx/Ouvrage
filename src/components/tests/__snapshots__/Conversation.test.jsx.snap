// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Conversation
    currentContact={
      Object {
        "avatar": "https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg",
        "firstName": "Manu",
        "id": "2",
        "lastName": "Patrois",
        "username": "epatrois",
      }
    }
    me={
      Object {
        "avatar": "https://cdn.pixabay.com/photo/2015/10/05/22/37/blank-profile-picture-973460_960_720.png",
        "birthdate": "838921463",
        "email": "maxencedewaele@gmail.com",
        "firstName": "Maxence",
        "id": "5",
        "lastName": "De Waele",
        "username": "mdewaele",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div>
        <header>
          <h2>
            Manu
             
            Patrois
          </h2>
          <img
            className="user-avatar"
            src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
          />
        </header>
        <main>
          <ul
            className="message-list"
          />
        </main>
        <footer
          className="message-section"
        >
          <input
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Votre message..."
            type="text"
            value=""
          />
        </footer>
      </div>,
      "className": "conversation",
      "onScroll": [Function],
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <header>
            <h2>
              Manu
               
              Patrois
            </h2>
            <img
              className="user-avatar"
              src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
            />
          </header>,
          <main>
            <ul
              className="message-list"
            />
          </main>,
          <footer
            className="message-section"
          >
            <input
              onChange={[Function]}
              onKeyPress={[Function]}
              placeholder="Votre message..."
              type="text"
              value=""
            />
          </footer>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h2>
                Manu
                 
                Patrois
              </h2>,
              <img
                className="user-avatar"
                src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "Manu",
                  " ",
                  "Patrois",
                ],
              },
              "ref": null,
              "rendered": Array [
                "Manu",
                " ",
                "Patrois",
              ],
              "type": "h2",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "user-avatar",
                "src": "https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
          ],
          "type": "header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ul
              className="message-list"
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "message-list",
            },
            "ref": null,
            "rendered": Array [],
            "type": "ul",
          },
          "type": "main",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              onChange={[Function]}
              onKeyPress={[Function]}
              placeholder="Votre message..."
              type="text"
              value=""
            />,
            "className": "message-section",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "onChange": [Function],
              "onKeyPress": [Function],
              "placeholder": "Votre message...",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "footer",
        },
      ],
      "type": "div",
    },
    "type": "section",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div>
          <header>
            <h2>
              Manu
               
              Patrois
            </h2>
            <img
              className="user-avatar"
              src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
            />
          </header>
          <main>
            <ul
              className="message-list"
            />
          </main>
          <footer
            className="message-section"
          >
            <input
              onChange={[Function]}
              onKeyPress={[Function]}
              placeholder="Votre message..."
              type="text"
              value=""
            />
          </footer>
        </div>,
        "className": "conversation",
        "onScroll": [Function],
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <header>
              <h2>
                Manu
                 
                Patrois
              </h2>
              <img
                className="user-avatar"
                src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
              />
            </header>,
            <main>
              <ul
                className="message-list"
              />
            </main>,
            <footer
              className="message-section"
            >
              <input
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder="Votre message..."
                type="text"
                value=""
              />
            </footer>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <h2>
                  Manu
                   
                  Patrois
                </h2>,
                <img
                  className="user-avatar"
                  src="https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg"
                />,
              ],
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": Array [
                    "Manu",
                    " ",
                    "Patrois",
                  ],
                },
                "ref": null,
                "rendered": Array [
                  "Manu",
                  " ",
                  "Patrois",
                ],
                "type": "h2",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "user-avatar",
                  "src": "https://pbs.twimg.com/profile_images/781247417819725825/g8Rk4ljG_400x400.jpg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
            ],
            "type": "header",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <ul
                className="message-list"
              />,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [],
                "className": "message-list",
              },
              "ref": null,
              "rendered": Array [],
              "type": "ul",
            },
            "type": "main",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                onChange={[Function]}
                onKeyPress={[Function]}
                placeholder="Votre message..."
                type="text"
                value=""
              />,
              "className": "message-section",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "onChange": [Function],
                "onKeyPress": [Function],
                "placeholder": "Votre message...",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "footer",
          },
        ],
        "type": "div",
      },
      "type": "section",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
